### ISSUE TITLE
Help with LibGDX on Ruboto

### ISSUE THREAD

**Issue by ashes999** [ruboto/ruboto#743]
**TLDR:** I don't have a lot of C++ or (raw) Android experience, and would like some help getting libGDX to work on Ruboto.

-----

`gosu-android` is an excellent gem, albeit one that is fairly new (not mature) and no longer under active development. I would like to integrate [libGDX](https://github.com/libgdx/libgdx), which is a mature, cross-platform, Java-based library -- albeit one that's about 50% C/C++ code.

You can see the current working state [here](https://github.com/ashes999/terrace/tree/master/template/android-libgdx).  Here's how I got this far:

- Create a "hello world" libGDX generated desktop/Android using the setup tool.
- Convert the code over to JRuby, and run it.
- Download the [latest libGDX release](http://libgdx.badlogicgames.com/releases/), which is 1.6.1, because the auto-generated setup JAR didn't give me the `-native` JARs that I needed.
- Copy/paste a working `gosu-android` Ruboto example.
- Convert the code from the libGDX generated Android launcher, and the main game, into Ruby. Place into `template_activity.rb`.

This almost seems to work. I no longer get errors around missing/undefined Java classes (like the `AndroidApplication` class).  Here's what my main app looks like:

```
require 'ruboto/widget'
require 'ruboto/util/toast'

# libGDX requires
java_import com.badlogic.gdx.ApplicationAdapter;
java_import com.badlogic.gdx.Gdx;
java_import com.badlogic.gdx.graphics.GL20;
java_import com.badlogic.gdx.graphics.Texture;
java_import com.badlogic.gdx.graphics.g2d.SpriteBatch;

# libGDX/Android requires
java_import android.os.Bundle;
java_import com.badlogic.gdx.backends.android.AndroidApplicationConfiguration;
java_import com.badlogic.gdx.backends.android.AndroidApplication;

class AndroidLauncher < AndroidApplication

  def onCreate(savedInstanceState)
    super.onCreate(savedInstanceState)
    config = AndroidApplicationConfiguration.new
    initialize(TestGame.new, config)
  end
end

class TestGame < ApplicationAdapter
  def create
    @batch = SpriteBatch.new
    @img = Texture.new("assets/badlogic.jpg")
  end

  def render
    Gdx.gl.glClearColor(1, 0, 0, 1)
    Gdx.gl.glClear(GL20.GL_COLOR_BUFFER_BIT)
    @batch.begin
    @batch.draw(@img, 0, 0)
    @batch.end
  end
end

class TemplateActivity
  def on_create(bundle)
    super(bundle)
    rescue Exception => e
      puts "#{ e } (#{ e.class } #{e.message} #{e.backtrace.inspect} )!"
  end

  def on_ready
    AndroidLauncher.new.onCreate(nil)
    rescue Exception => e
      puts "#{ e } (#{ e.class } #{e.message} #{e.backtrace.inspect} )!"
  end
end
```

Under the `libs` directory (which comes with `bundler.jar`), I've added the following:

- `armeabi`, `armeabi-v7a`, and `x86` folders (from libGDX's Android project; these contain `libgdx.so` and `libgdx-box2d.so`).
- `gdx.jar`
- `gdx-backend-android.jar`
- `gdx-natives.jar`
- `gdx-box2d.jar`
- `gdx.box2d-natives.jar`

I'm not sure if the `.so` files belong here. When I run the app, I see this:

```
D/dalvikvm( 2073): Trying to load lib /data/data/com.terrace.template/lib/libgdx.so 0xa69eba58
D/dalvikvm( 2073): Added shared lib /data/data/com.terrace.template/lib/libgdx.so 0xa69eba58
D/dalvikvm( 2073): No JNI_OnLoad found in /data/data/com.terrace.template/lib/libgdx.so 0xa69eba58, skipping init
```

Running the app gives me an empty screen (not a crash). Immediately following the above logs, I see a stack dump:

```
D/RUBOTO  ( 2073): Script load took 1431ms
D/RUBOTO  ( 2073): EntryPointActivity onCreate:
I/Choreographer( 2073): Skipped 100 frames!  The application may be doing too much work on its main thread.
W/EGL_genymotion( 2073): eglSurfaceAttrib not implemented
D/OpenGLRenderer( 2073): TextureCache::get: create texture(0xb987af20): name, size, mSize = 114, 2304, 1084672
D/OpenGLRenderer( 2073): TextureCache::get: create texture(0xb98c2858): name, size, mSize = 116, 9216, 1093888
I/ActivityManager(  297): Displayed com.terrace.template/.TemplateActivity: +1s748ms
D/RUBOTO  ( 2073): SplashActivity onDestroy: 
E/WindowManager( 2073): Activity org.ruboto.SplashActivity has leaked window com.android.internal.policy.impl.PhoneWindow$DecorView@534ffe64 that was originally added here
E/WindowManager( 2073): android.view.WindowLeaked: Activity org.ruboto.SplashActivity has leaked window com.android.internal.policy.impl.PhoneWindow$DecorView@534ffe64 that was originally added here
E/WindowManager( 2073): 	at android.view.ViewRootImpl.<init>(ViewRootImpl.java:374)
E/WindowManager( 2073): 	at android.view.WindowManagerImpl.addView(WindowManagerImpl.java:292)
E/WindowManager( 2073): 	at android.view.WindowManagerImpl.addView(WindowManagerImpl.java:224)
E/WindowManager( 2073): 	at android.view.WindowManagerImpl$CompatModeWrapper.addView(WindowManagerImpl.java:149)
E/WindowManager( 2073): 	at android.view.Window$LocalWindowManager.addView(Window.java:547)
E/WindowManager( 2073): 	at android.app.Dialog.show(Dialog.java:277)
E/WindowManager( 2073): 	at android.app.ProgressDialog.show(ProgressDialog.java:116)
E/WindowManager( 2073): 	at android.app.ProgressDialog.show(ProgressDialog.java:104)
E/WindowManager( 2073): 	at org.ruboto.SplashActivity.showProgress(SplashActivity.java:171)
E/WindowManager( 2073): 	at org.ruboto.SplashActivity.initJRuby(SplashActivity.java:75)
E/WindowManager( 2073): 	at org.ruboto.SplashActivity.onCreate(SplashActivity.java:43)
E/WindowManager( 2073): 	at android.app.Activity.performCreate(Activity.java:5008)
E/WindowManager( 2073): 	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1079)
E/WindowManager( 2073): 	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2023)
E/WindowManager( 2073): 	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2084)
E/WindowManager( 2073): 	at android.app.ActivityThread.access$600(ActivityThread.java:130)
E/WindowManager( 2073): 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1195)
E/WindowManager( 2073): 	at android.os.Handler.dispatchMessage(Handler.java:99)
E/WindowManager( 2073): 	at android.os.Looper.loop(Looper.java:137)
E/WindowManager( 2073): 	at android.app.ActivityThread.main(ActivityThread.java:4745)
E/WindowManager( 2073): 	at java.lang.reflect.Method.invokeNative(Native Method)
E/WindowManager( 2073): 	at java.lang.reflect.Method.invoke(Method.java:511)
E/WindowManager( 2073): 	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:786)
E/WindowManager( 2073): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:553)
E/WindowManager( 2073): 	at dalvik.system.NativeStart.main(Native Method)
```

The main issue I have is that I'm not sure how to load my app.  Here's what the raw Java code looks like for the Android launcher (sans imports):

```
public class AndroidLauncher extends AndroidApplication {
	@Override
	protected void onCreate (Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		AndroidApplicationConfiguration config = new AndroidApplicationConfiguration();
		initialize(new TerraceGame(), config);
	}
}
```

I do have my own `AndroidLauncher` class. `AndroidApplication` [seems to extend `Activity`](https://github.com/libgdx/libgdx/blob/fa21bbd665de1d25f326bc0f78acc862e6ca2a7b/backends/gdx-backend-android/src/com/badlogic/gdx/backends/android/AndroidApplication.java).

What I lack is knowledge of how to launch it. This line from `on_ready` doesn't seem right:

`AndroidLauncher.new.onCreate(nil)`

Any help would be greatly appreciated.  As I mentioned up above, you can grab the full repository [here](https://github.com/ashes999/terrace/tree/master/template/android-libgdx) and build it yourself (`rake debug` does the job).

I'm testing on Genymotion with an Android 4.1 VM, if that matters.

**Comment by donv** [comment]
Hi @ashes999 !

This looks like a really interesting project!  I would love to help, and others on the project will surely help as well.

I have not looked at the repository yet (too much todo :disappointed:), but I have a couple of remarks right now:

I have not used any native libraries yet, but you should put them in the ```lib``` directory.  To automatically detect the architecture of the library, you should put it under the a subdirectory with the architecture name like ```lib/armeabi/libfoo.so```.  Note that you need one version of the library for every platform you want the app to run on (armeabi, x86, x86_64, etc.).

The "Activity org.ruboto.SplashActivity has leaked window" stack trace you see is probably not related, and it is investigated here: https://github.com/ruboto/ruboto/issues/630 .

I hope this helps, and please keep at it and report your progress on the mailing list or on the IRC or Gitter channel.

I have forked the project here:  https://github.com/ruboto/terrace and I will take a look.  Others on the Ruboto project are also welcome to take a look at see what they can find.  It should be a good activity to get into Android and Ruboto.

**Comment by donv** [comment]
Hi @ashes999 !

I have taken a look at the project and the immediate thing is that you don't display any content :smile: 

In ```TemplateActivity#onCreate``` you should set up some visual content.  I have not looked at terrace yet, so I don't know how you should draw content, but the blank screen you see at startup is correct since you don't actually add any content to the screen.

Content is added using the ```self.content_view =```,  ```self.set_content_view(view)``` or ```self.setContentView(view)``` methods.

Example:  https://github.com/ruboto/ruboto/blob/master/test/activity/image_button_activity.rb#L17

**Comment by ashes999** [comment]
Hi,

I'm not sure I understand Android well enough. I know that you need an activity of some sort to start your app. In my case, I do have `class TemplateActivity`. 

There's nothing in `on_create`, but in `on_ready` ([this line](https://github.com/ashes999/terrace/blob/master/template/android-libgdx/src/template_activity.rb#L54)), I have `AndroidLauncher.new.onCreate(nil)`. This is almost certainly incorrect.

I think I don't need to add content per-se; libGDX has it's own way of launching the app, which I believe is just to set up the startup activity correctly.  You can read about it [here](https://github.com/libgdx/libgdx/wiki/A-simple-game#configuring-the-starter-classes). It's basically:

- A subclass of `AndroidActivity` that calls `initialize(gameClassInstance, config)`
- A game class that extends `ApplicationAdapter` (`ApplicationListener` is the old one, I think)

I'm not sure what I'm missing. If your claim is that the window leaking is a red herring, I'm probably missing something on the libGDX side, not the Ruboto side. But what? How can I figure this out?

**Comment by donv** [comment]
Hi @ashes999 !

```on_ready``` is not an Android Activity lifecycle method.  ```onResume``` is maybe what you were looking for.

I have added a pull request that should help you a bit along.  I am still getting a NullPointerException inside the Java classes, though.  Maybe you can figure that out :smile: .

**Comment by ashes999** [comment]
Thanks @donv , I'll take a look. 

I used `on_ready` because this is what I had working with `Gosu-Android`

**Comment by donv** [comment]
Yeah, ```on_ready``` is a Gosu lifecycle method, I think.  @ashes999 please report your progress, and I hope you can merge my pull request.

**Comment by ashes999** [comment]
It's progress, but it crashes, I'm not sure why. I'm going to merge your PR, and leave this as-is for now to revisit later when I have more time.

Thanks for your help!

**Comment by donv** [comment]
Glad to help.  Looking forward to hearing from you again.

**Comment by ashes999** [comment]
I asked for help on the libGDX side but didn't get any useful feedback. You can read lots of details on [GameDev.SE](http://gamedev.stackexchange.com/questions/104399/nullpointerexception-in-android-app-activity-oncreateandroid-app-activity-java) (I also posted on the libGDX forums).

The exception I'm getting now is in the `super` call in `onCreate`. It seems like something in the Android Activity class is unexpectedly null:

```
D/dalvikvm( 1989): DEX prep '/data/data/com.terrace.template/app_dex/org/jruby/proxy/com/badlogic/gdx/backends/android/AndroidApplication$Proxy0.jar': unzip in 0ms, rewrite 10ms
I/System.out( 1989): Exception starting AndroidLauncher:  (Java::JavaLang::NullPointerException )
I/System.out( 1989): android.app.Activity.onCreate(android/app/Activity.java:874)
I/System.out( 1989): java.lang.reflect.Method.invokeNative(Native Method)
I/System.out( 1989): java.lang.reflect.Method.invoke(java/lang/reflect/Method.java:511)
I/System.out( 1989): RUBY.onCreate(jar:file:/data/app/com.terrace.template-1.apk!/template_activity.rb:35)
```

I couldn't figure out how to debug this. I know the vanilla (generated) libGDX project for Android works (as does the libGDX desktop project with JRuby); there must be some difference between the two Android projects, but where? `AndroidManifest.xml` and `local.properties` seem similar enough.

Any thoughts @donv ?

**Comment by ashes999** [comment]
I've been thinking about this some more. I think the issue might be related to how libGDX and Ruboto initialize. Some googling shows that this exception usually occurs when you didn't specify, in your `Intent` class, what the next class should be.

I added some trace statements. It seems that, in Ruboto, the subsequent class is set. And I've updated the `AndroidManifest.xml` settings to match the libGDX ones, especially on the intent.

@donv I don't know much about Android initialization. Could you take a look and see if you can figure out what's causing the null pointer exception?

Alternatively, I can try doing this with 100% Java classes, and just using Ruboto as a thin shell; that would rule out the possibility that it's something in the Ruby versions of the class that are the problem.

**Comment by ashes999** [comment]
Many, many thanks @donv -- I kept coming back to your initialization changes, which formed a solid base for me to figure this out.

For future readers: I got this working. You can see the current, working version in my [libgdx/ruboto repository](https://github.com/ashes999/libgdx-ruboto).

Basically, the architecture is:

- Main game class is a pure Ruby class
- `AndroidLauncher` and all other classes are libGDX/Java classes
- I create the main game class, pass it to the Java side, and run it through a Java activity

Thanks again for all your help.
